convex/lp/simplex.go

```diff
@@ -528,44 +534,53 @@ func findInitialBasic(A mat64.Matrix, b []float64) ([]int, *mat64.Dense, []float
 	c := make([]float64, n+1)
 	c[n] = 1
 
-	// Solve the Phase 2 linear program.
+	// Solve the Phase I linear program.
 	_, xOpt, newBasic, err := simplex(basicIdxs, c, aNew, b, 1e-10)
 	if err != nil {
 		return nil, nil, nil, errors.New(fmt.Sprintf("lp: error finding feasible basis: %s", err))
 	}
 
-	// If n+1 is part of the solution basis then the problem is infeasible. If
-	// not, then the problem is feasible and newBasic is an initial feasible
-	// solution.
+	// The original LP is infeasible if the added variable has non-zero value
+	// in the optimal solution to the Phase I problem.
 	if math.Abs(xOpt[n]) > phaseIZeroTol {
 		return nil, nil, nil, ErrInfeasible
 	}
 
-	// The value is zero. First, see if it's not in the basis (feasible solution).
-	basicIdx := -1
-	basicMap := make(map[int]struct{})
+	// The basis found in Phase I is a feasible solution to the original LP if
+	// the added variable is not in the basis.
+	addedIdx := -1
 	for i, v := range newBasic {
 		if v == n {
-			basicIdx = i
+			addedIdx = i
 		}
-		basicMap[v] = struct{}{}
 		xb[i] = xOpt[v]
 	}
-	if basicIdx == -1 {
-		// Not in the basis.
+	if addedIdx == -1 {
 		extractColumns(ab, A, newBasic)
 		return newBasic, ab, xb, nil
 	}
 
-	// The value is zero, but it's in the basis. See if swapping in another column
-	// finds a feasible solution.
+	// The value of the added variable is in the basis, but it has a zero value.
+	// See if exchanging another variable into the basic set finds a feasible
+	// solution.
+	basicMap := make(map[int]struct{})
+	for _, v := range newBasic {
+		basicMap[v] = struct{}{}
+	}
+	var set bool
 	for i := range xOpt {
 		if _, inBasic := basicMap[i]; inBasic {
 			continue
 		}
-		newBasic[basicIdx] = i
-		extractColumns(ab, A, newBasic)
-		xb, err := initializeFromBasic(ab, b)
+		newBasic[addedIdx] = i
+		if set {
```

Comment:
Actually, I accidentally nuked my local version of this... I'll submit this in the next PR